lib/vcs.ml

```diff
@@ -64,28 +64,49 @@ let find_git () =
       | dir, (_, `Exited 0) -> Ok (Some (v `Git git ~dir:Fpath.(v dir)))
       | _ -> Ok None )
 
-let cmd_error cmd = function
-  | `Exited c ->
-      R.error_msgf "The following command exited with code %d:@\n%a" c Cmd.pp
-        cmd
-  | `Signaled c ->
-      R.error_msgf "The following command exited with signal %d:@\n%a" c Cmd.pp
-        cmd
+let cmd_error cmd ?err_msg =
+  match err_msg with
+  | None | Some "" -> (
+      function
+      | `Exited c ->
+          R.error_msgf "The following command exited with code %d:@\n%a" c
+            Cmd.pp cmd
+      | `Signaled c ->
+          R.error_msgf "The following command exited with signal %d:@\n%a" c
+            Cmd.pp cmd )
+  | Some err_msg -> (
+      function
+      | `Exited c ->
+          R.error_msgf "Exit code %d from command @\n`%a`:@\n%s" c Cmd.pp cmd
+            err_msg
+      | `Signaled c ->
+          R.error_msgf "Signal %d from command @\n`%a`:@\n%s" c Cmd.pp cmd
+            err_msg )
+
+module D = struct
+  let string = Sos.out ""
+
+  let list = Sos.out []
+
+  let unit = Sos.out ()
+end
 
 let run_git ~dry_run ?force ~default r args out =
   let git = Cmd.(cmd r %% args) in
-  Sos.run_out ~dry_run ~sandbox:false ?force git ~err:OS.Cmd.err_null ~default
-    out
-  >>= function
-  | v, (_, `Exited 0) -> Ok v
-  | _, (_, st) -> cmd_error git st
+  Sos.run_out_err ~dry_run ~sandbox:false ?force git ~default out
+  >>= fun response ->
+  match response.status with
+  | `Exited 0 -> Ok response.output
+  | _ -> cmd_error git ~err_msg:response.err_msg response.status
 
 let git_is_dirty r =
   let status = Cmd.(cmd r %% git_work_tree r % "status" % "--porcelain") in
-  OS.Cmd.(run_out ~err:err_null status |> out_string) >>= function
-  | "", (_, `Exited 0) -> Ok false
-  | _, (_, `Exited 0) -> Ok true
-  | _, (_, st) -> cmd_error status st
+  Sos.run_out_err ~dry_run:false status ~default:D.string OS.Cmd.out_string
+  >>= fun response ->
+  match (response.output, response.status) with
+  | "", `Exited 0 -> Ok false
+  | _, `Exited 0 -> Ok true
+  | _, _ -> cmd_error status ~err_msg:response.err_msg response.status
```

Comment:
This is just a style note but you could have written it like this:
```suggestion
  Sos.run_out_err ~dry_run:false status ~default:D.string OS.Cmd.out_string
  >>= function
  | {output = ""; status = `Exited 0; _} -> Ok false
  | {status = `Exited 0; _} -> Ok true
  | {err_msg; status; _} -> cmd_error status ~err_msg status
```

Reply:
Oh, nice. I remember I read about record pattern matching, thought it was cool and then forgot it again. Thanks!